### PR TITLE
#144: Extracted load_models into seperate class

### LINKED ISSUE #144: Extract base_model_udf.load_models into separate class
# Background:

- Currently, we use from_pretrained with the cache_dir option to load the model and upload the cache directory to the BucketFS
- The current approach in general works, but seems to check each time online if the model exists,  we have access and  we have the newest version
- This is not possible in some deployments 
- To solve this issue, we could follow the following [example](https://huggingface.co/docs/transformers/v4.32.1/en/installation#fetch-models-and-tokenizers-to-use-offline) using save_pretrained which stores the model into a directory which we then could upload
- We can then load the model with from_pretrained from a local path
- It would be easier to change the model loading once it is extracted into its own class

# Acceptance Criteria

- [ ]  Extract [base_model_udf.load_models](https://github.com/exasol/transformers-extension/blob/1321af15f1cf0e9ac68ca12c02a3d6e4318a2cb1/exasol_transformers_extension/udfs/models/base_model_udf.py#L187C13-L187C13) into seperate class and inject it into base_model_udf

### PR DESCRIPTION
**Pull request by MarleneKress79789**
### All Submissions:

* [x] Is the title of the Pull Request correct?
* [x] Is the title of the corresponding issue correct?
* [x] Have you updated the changelog?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../../pulls) for the same update/change?  
* [x] Are you mentioning the issue which this PullRequest fixes ("Fixes...")
* [x] Before you merge don't forget to run tests in AWS CodeBuild, by adding `[CodeBuild]` to the commit message

Fixes #144

### CODE REVIEW

File: exasol_transformers_extension/udfs/models/base_model_udf.py [hunk 1]
```diff
@@ -199,36 +215,10 @@ def clear_device_memory(self):
         """
         Delete models and free device memory
         """
-        self.last_loaded_model = None
-        self.last_loaded_tokenizer = None
+        self.model_loader.last_loaded_model = None
+        self.model_loader.last_loaded_tokenizer = None
```
Comment (ahsimb) [inline exasol_transformers_extension/udfs/models/base_model_udf.py:219]: I would add a `clear` method to the ModelLoader setting these two variables to None.
If `last_laded_model` and `last_loaded_tokenizer` still need to be visible I would make them properties.

BTW, if the `last_created_pipeline` keeps references to these objects they won't be garbage-collected. So you probably need to it None too.

File: tests/unit_tests/udfs/base_model_dummy_implementation.py [hunk 2]
```diff
@@ -6,6 +6,38 @@
...
+                    cache_dir,
+                    token_conn_obj) -> None:
+        token = False
+        self.last_loaded_model = self.base_model.from_pretrained(
+            model_name, cache_dir=cache_dir, use_auth_token=token)
+        self.last_loaded_tokenizer = self.tokenizer.from_pretrained(
+            model_name, cache_dir=cache_dir, use_auth_token=token)
+        return None
+
+
```
Comment (ahsimb) [inline tests/unit_tests/udfs/base_model_dummy_implementation.py:40]: Why can't you use the actual ModelLoader with a pipeline-like function that does nothing and returns None?

File: exasol_transformers_extension/udfs/models/base_model_udf.py (hunk 1 above)
Comment (MarleneKress79789) [reply exasol_transformers_extension/udfs/models/base_model_udf.py:219]: will move the clear method from base udf also

File: exasol_transformers_extension/utils/load_model.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        self.task_name = task_name
+        self.device = device
+        self.last_loaded_model = None
+        self.last_loaded_tokenizer = None
+        self.last_loaded_model_key = None
+
+    def load_models(self, model_name: str,
+                    current_model_key,
+                    cache_dir,
+                    token_conn_obj) -> None:
```
[the remark this answers is not in the document]
Comment (MarleneKress79789) [reply exasol_transformers_extension/utils/load_model.py:23]: regarding the documentation strings and stuff: this whole class will be replaced with the one that will be created in #145. so i would rather spend the time to do proper docu for the new class and leave this one as is.

File: tests/unit_tests/udfs/base_model_dummy_implementation.py (hunk 2 above)
Comment (MarleneKress79789) [reply tests/unit_tests/udfs/base_model_dummy_implementation.py]: You are completely right, i misunderstood. changed

File: exasol_transformers_extension/utils/load_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahsimb) [reply exasol_transformers_extension/utils/load_model.py:23]: I would say the released code should be in a state of completeness. If the class morphs into something else so will the documentation. Plus it doesn't take long to get the docstrings sorted.

**Review by ahsimb — APPROVED** [review]

File: exasol_transformers_extension/utils/load_model.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        Load model and tokenizer model from the cached location in bucketfs.
+        If the desired model is not cached, this method will attempt to
+        download the model to the read-only path /bucket/.. and cause an error.
+        This error will be addressed in ticket
+        https://github.com/exasol/transformers-extension/issues/43.
+
+        :param model_name: The model name to be loaded
+        """
+        token = False
+        if token_conn_obj:
```
[the remark this answers is not in the document]
Comment (MarleneKress79789) [reply exasol_transformers_extension/utils/load_model.py:34]: new bug ticket added here: https://github.com/exasol/transformers-extension/issues/163

**Pull request merged**